opm/simulators/linalg/FlexibleSolver_impl.hpp

```diff
@@ -141,10 +155,22 @@ namespace Dune
         // Sequential case.
         linearoperator_for_solver_ = &op;
         auto child = prm.get_child_optional("preconditioner");
-        preconditioner_ = Opm::PreconditionerFactory<Operator,Dune::Amg::SequentialInformation>::create(op,
-                                                                       child ? *child : Opm::PropertyTree(),
-                                                                       weightsCalculator,
-                                                                       pressureIndex);
+        const std::string& type = child ? child->get<std::string>("type","ParOverILU0"):std::string("ParOverILU0");
+        if (type != "amg") {
+            preconditioner_ = Opm::PreconditionerFactory<Operator,Dune::Amg::SequentialInformation>::create(op,
+                                                                           child ? *child : Opm::PropertyTree(),
+                                                                           weightsCalculator,
+                                                                           pressureIndex);
+        } else {
+            using RangeType = VectorType; // Assuming symmetry: domain == range
+            using SOT = Dune::MatrixAdapter<MatrixType, VectorType, RangeType>;
+            auto op_prec = std::make_shared<SOT>(op.getmat());
+            preconditioner_ = Opm::PreconditionerFactory<SOT,Dune::Amg::SequentialInformation>::create(*op_prec,
```

Comment:
I think this is not going to be correct, but it is a bit tricky so I may be mistaken. The code change in ISTLSolver.cpp should be sufficient I think. I believe this will force the no-well operator when requesting the "amg" preconditioner, I do not see the reason why? Same with the parallel case above, but it is more obviously suspicious when there is a change to the sequential case for a parallel-only improvement!

Reply:
Thanks! Yes, that is in a separate [commit and branch](https://github.com/lisajulia/opm-simulators/tree/ilu-op-in-amg-force-ghostlastmatrixoperator-for-comparison) now, Andreas explained that he initially had added this to do a fair comparion between AMG and CPR.